moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp

```diff
@@ -156,16 +156,78 @@ bool ChompPlanner::solve(const planning_scene::PlanningSceneConstPtr& planning_s
 
   ros::WallTime create_time = ros::WallTime::now();
 
-  ChompOptimizer optimizer(&trajectory, planning_scene, req.group_name, &params, start_state);
-  if (!optimizer.isInitialized())
+  int replan_count = 0;
+  bool replan_flag = false;
+  double org_learning_rate = 0.04, org_ridge_factor = 0.0, org_planning_time_limit = 10;
+  int org_max_iterations = 200;
+
+  // storing the initial chomp parameters values
+  org_learning_rate = params.learning_rate_;
+  org_ridge_factor = params.ridge_factor_;
+  org_planning_time_limit = params.planning_time_limit_;
+  org_max_iterations = params.max_iterations_;
+
+  ChompOptimizer* optimizer;
+
+  // create a non_const_params variable which stores the non constant version of the const params variable
+  ChompParameters non_constant = non_constant.getNonConstantParams(params);
+
+  // do while loop for replanning (recovery behaviour) if collision free optimized solution not found
+  do
   {
-    ROS_ERROR_STREAM_NAMED("chomp_planner", "Could not initialize optimizer");
-    res.error_code.val = moveit_msgs::MoveItErrorCodes::PLANNING_FAILED;
-    return false;
-  }
-  ROS_DEBUG_NAMED("chomp_planner", "Optimization took %f sec to create", (ros::WallTime::now() - create_time).toSec());
+    if (replan_flag)
+    {
+      non_constant.learning_rate_ += 0.02;     // increase learning rate in hope to find a successful path
+      non_constant.ridge_factor_ += 0.002;     // increase ridge factor to avoid obstacles
+      non_constant.planning_time_limit_ += 5;  // add 5 additional secs in hope to find a solution
+      non_constant.max_iterations_ += 50;      // increase maximum iterations
+    }
+
+    optimizer = new ChompOptimizer(&trajectory, planning_scene, req.group_name, &non_constant, start_state);
```

Comment:
this is a memory leak - ``optimizer`` is never deleted. can you refactor this code to instantiate the ChompOptimizer only once, but use a``setParams()`` function to change the params for each loop?

Reply:
thanks, I didn't realize this earlier. I am currently [deleting the optimizer](https://github.com/ros-planning/moveit/pull/987/files#diff-678f73e93bab1f4e5438e0afbb3cc56bR216) before creating a new ChompOptimizer with new parameters, if ChompOptimizer is instantiated only once its, difficult to use the [`const ChompParameters parameters_`](https://github.com/raghavendersahdev/moveit/blob/kinetic-devel/moveit_planners/chomp/chomp_motion_planner/include/chomp_motion_planner/chomp_optimizer.h#L133) as it is used at a lot of places in the [`optimize`](https://github.com/raghavendersahdev/moveit/blob/kinetic-devel/moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp#L318) method, using a non const version would not be good here, coz I need to do too many changes in optimize method. So I simply delete the pointer here to prevent a memory leak.